pkg/internal/controller/controller.go

```diff
@@ -98,9 +96,7 @@ type Controller struct {
 
 // watchDescription contains all the information necessary to start a watch.
 type watchDescription struct {
```

Comment:
So I actually realized I made a mistake with how I designed it with having methods in the internalSource to enable configuring a metadata watch after the fact: We assert the incoming type to the sources generic type and that will completely fail if object is a metadata object. So In re-designed this to not store a source at all, but all parts to construct a source in the `watchDescription`, that way we do not have to manipulate it after constructing it and IMHO its much cleaner.

Reply:
There is a big disclaimer on watchDescription that it is not supposed to be used alone, and that the `PartialObjectMetadata` should be used. I believe that this field even does not make any effect, so I don’t think it should be carried over into new implementations. Generics will work even better with `PartialObjectMetadata` passed. I mentioned this on the event handler PR https://github.com/kubernetes-sigs/controller-runtime/pull/2698#issue-2160848733

Reply:
I am realizing in the meantime it coincidentally worked, because client.Object is implemented by both the metaobject and any type, so the assertion actually succeeded.

Reply:
I actually got rid of the `watchDescription` in favor of building `source.Source` via changed interfaces before starting it. It can be used like:
```go
var src source.PrepareSyncing
var prepared source.SyncingSource
prepared = src.Prepare(handler, predicates…)
```
Then you can do something like
```go
c.Watch(source.Kind(mgr.GetCache(), &appsv1.ReplicaSet{}).Prepare(&handler.EnqueueRequestForObject{}))
```
I opened a draft #2784 to compare implementations, also did a test rewrite [here](https://github.com/Danil-Grigorev/cluster-api/pull/1) (to see how it feels) for cluster-api, as it has lots of predicates and map functions where generics will be the most useful.

Reply:
This PR avoids that entire issue by moving all args into source construction. IMHO that is a lot simpler than requiring sources to implement a `Prepare`.

Reply:
@Danil-Grigorev your approach requires something like this:
```
builder.Add(builder.For(b,
		&bootstrapv1.KubeadmConfig{},
		predicates.ResourceNotPausedAndHasFilterLabel(ctrl.LoggerFrom(ctx), r.WatchFilterValue, &bootstrapv1.KubeadmConfig{}),
	)).
```

IMHO something like that is a huge no-no:
* We break every single user of controller-runtime bigtimes
* It is super annoying to have to have the custom wrappers for methods like `For` etc, especially in contexts where none of this matters like you just want to enqueue the object, its owner etc - Which is the vast majority use case

Reply:
But you can still use all of the existing methods of the builder in a non-generic context, I don’t see how is this breaking. And additional wrappers allow you to typecheck all your event handlers and predicates on the compiler level within the method invocation. Which is extremely useful if you go beyond 1-2 predicate methods and do complex logic with And/Or. I noticed this while doing CAPI code rewrite. Adapters are mainly made for scenarios when you explicitly don’t want to rewrite your code to generics (like readability concerns). I would love to get rid of the builder requirement in these wrappers, it is only needed for passing validation (have For/Owns used at least once), otherwise passing a Manager would suffice. It is after all always available in `SetupWithManager` context.

Reply:
@Danil-Grigorev okay I missed that this is optional. But even then, the approach seems more complicated, you need a wraper to build a source. Why not just change things so that source construction itself takes all required args rather than needing an additional constructor func on top? I can see that it might be nice to optionally have the builder `For` and `Own` but that is it. Also, there is btw already `buillder.WatchesRawSource` of which your `Add` is a duplicate

Reply:
`Source` interface is being weakened by the rewrite of the start method. I’d like to preserve source requirements to pass predicates and a handler, maybe even just by a composite interface embedding updated `Source`. You are right, `WatchesRawSource` is duplicate of `Add`, but this allows for both to be stored in a single array without type erasure. Here is also an example for `builder.Watches` where every predicate or map method should always route to clusterv1.Cluster. This removes margin for error:
```go
Add(builder.Watches(mgr,
	&clusterv1.Cluster{},
	handler.EnqueueRequestsFromObjectMap(r.ClusterToKubeadmControlPlane),
	predicates.ResourceNotPausedAndHasFilterLabel(ctrl.LoggerFrom(ctx), r.WatchFilterValue, &clusterv1.Cluster{}),
	predicates.ClusterUnpausedAndInfrastructureReady(ctrl.LoggerFrom(ctx)),
)
```

My original motivation to go for a more complicated way was the hope to have as little breaking changes as possible, and yet allow for optional use of generics as a sort of first-class feature.